plutus-core/testlib/PlutusCore/Generators/PIR.hs

```diff
@@ -0,0 +1,1752 @@
+
+-- | Generate a well-kinded term in a given context
+genTypeWithCtx :: Map TyName (Kind ()) -> Kind () -> Gen (Type TyName DefaultUni ())
+genTypeWithCtx ctx k = runGenTm $ local (\ e -> e { geTypes = ctx }) (genType k)
+
+-- | Generate a well-kinded term in a given context
+genKindAndTypeWithCtx :: Map TyName (Kind ()) -> Gen (Kind(), Type TyName DefaultUni ())
+genKindAndTypeWithCtx ctx = do
+  k <- arbitrary
+  runGenTm $ local (\ e -> e { geTypes = ctx }) ((k,) <$> genType k)
+
+-- CODE REVIEW: does this exist anywhere??
+-- | `substClosedType x sub ty` substitutes the closed type `sub` for `x` in `ty`
+substClosedType :: TyName -> Type TyName DefaultUni () -> Type TyName DefaultUni () -> Type TyName DefaultUni ()
+substClosedType x sub ty =
+  case ty of
+    TyVar _ y
+      | x == y    -> sub
+      | otherwise -> ty
+    TyFun _ a b   -> TyFun () (substClosedType x sub a) (substClosedType x sub b)
+    TyApp _ a b   -> TyApp () (substClosedType x sub a) (substClosedType x sub b)
+    TyLam _ y k b
+      | x == y    -> ty
+      | otherwise -> TyLam () y k $ substClosedType x sub b
+    TyForall _ y k b
+      | x == y    -> ty
+      | otherwise -> TyForall () y k $ substClosedType x sub b
+    TyBuiltin{}   -> ty
+    TyIFix{}      -> ty
+
+-- CODE REVIEW: does this exist anywhere?
+builtinKind :: SomeTypeIn DefaultUni -> Kind ()
+builtinKind (SomeTypeIn t) = case t of
+  DefaultUniProtoList -> Star :-> Star
+  DefaultUniProtoPair -> Star :-> Star :-> Star
+  DefaultUniApply f _ -> let _ :-> k = builtinKind (SomeTypeIn f) in k
+  _                   -> Star
+
+-- * Shrinking types and kinds
+
+-- | Shriking-order on kinds
+leKind :: Kind () -> Kind () -> Bool
+leKind k1 k2 = go (reverse $ args k1) (reverse $ args k2)
+  where
+    args Type{}            = []
+    args (KindArrow _ a b) = a : args b
+
+    go [] _                = True
+    go _ []                = False
+    go (k : ks) (k' : ks')
+      | leKind k k' = go ks ks'
+      | otherwise   = go (k : ks) ks'
+
+-- | Strict shrinking order on kinds
+ltKind :: Kind () -> Kind () -> Bool
+ltKind k k' = k /= k' && leKind k k'
+
+-- | Take a type in a context and a new target kind
+--   Precondition: new kind is smaller or equal to old kind of the type.
+--   TODO (later): also allow changing which context it's valid in
+fixKind :: HasCallStack
+        => Map TyName (Kind ())
+        -> Type TyName DefaultUni ()
+        -> Kind ()
+        -> Type TyName DefaultUni ()
+fixKind ctx ty k
+  -- Nothing to do if we already have the right kind
+  | inferKind_ ctx ty == k = ty
+  | not $ k `leKind` inferKind_ ctx ty =
+      error "fixKind not smaller"
+  | otherwise = case ty of
+    -- Switch a variable out for a different variable of the right kind
+    TyVar _ _ | y : _ <- [ y | (y, k') <- Map.toList ctx
+                             , k == k' ] -> TyVar () y
+              | otherwise -> minimalType k
+    -- Try to fix application by fixing the function
```

Comment:
You kind-of want shrinking to be deterministic... You could argue that fixup should return a list of fixes though.

Reply:
TBH I wasn't really thinking about that, I was just thinking that this looks a lot like a generator. In fact, I don't really know why `shrink` returns a `[a]` instead of a `Gen a`. `Gen` is still deterministic, and returning a list is a poor man's nondeterminism. But I'll stop trying to redesign quickcheck :laughing:

Reply:
One reason you don't want `Gen` is that you may want to control the order: do the big shrinking steps first!

Reply:
@michaelpj 

> why not fix the argument?

It's impossible. You can't change the kind of `f x` by changing `x` (since we don't have any polymorphism and implicit arguments in the type language of Plutus Core) -- only by changing `f` (possibly together with `x`, though, but I don't think it's too important to worry about that).

> You could argue that fixup should return a list of fixes though.

Sounds nice, created PLT-874, so that we don't forget to try it out in future.